daemon/endpoint.go

```diff
@@ -48,50 +49,47 @@ func (h *getEndpoint) Handle(params GetEndpointParams) middleware.Responder {
 	metrics.SetTSValue(metrics.EventTSAPI, time.Now())
 	log.WithField(logfields.Params, logfields.Repr(params)).Debug("GET /endpoint request")
 
-	modelChan := make(chan *models.Endpoint, 1)
-	defer close(modelChan)
+	var (
+		epModelsWg, epsAppendWg sync.WaitGroup
+		convertedLabels         labels.Labels
+		resEPs                  []*models.Endpoint
+	)
 
-	// FIXME @aanm hide use of endpoint manager lock
-	endpointmanager.Mutex.RLock()
-	numEndpoints := len(endpointmanager.Endpoints)
-	eps := make([]*models.Endpoint, 0, numEndpoints)
-
-	if params.Labels == nil {
-		for _, v := range endpointmanager.Endpoints {
-			go func(ep *endpoint.Endpoint) {
-				modelChan <- ep.GetModel()
-			}(v)
-		}
-		endpointmanager.Mutex.RUnlock()
-		for i := 0; i < numEndpoints; i++ {
-			eps = append(eps, <-modelChan)
-		}
-		return NewGetEndpointOK().WithPayload(eps)
+	if params.Labels != nil {
+		// Convert params.Labels to model that we can compare with the endpoint's labels.
+		convertedLabels = labels.NewLabelsFromModel(params.Labels)
 	}
 
-	// Convert params.Labels to model that we can compare with the endpoint's labels.
-	convertedLabels := labels.NewLabelsFromModel(params.Labels)
+	eps := endpointmanager.GetEndpoints()
+	epModelsCh := make(chan *models.Endpoint, len(eps))
 
-	for _, v := range endpointmanager.Endpoints {
-		go func(ep *endpoint.Endpoint) {
-			ep.Mutex.RLock()
+	epModelsWg.Add(len(eps))
+	for k := range eps {
+		go func(wg *sync.WaitGroup, epChan chan<- *models.Endpoint, ep *endpoint.Endpoint) {
 			if ep.HasLabels(convertedLabels) {
-				modelChan <- ep.GetModelRLocked()
+				epChan <- ep.GetModel()
 			}
-			ep.Mutex.RUnlock()
-		}(v)
+			wg.Done()
+		}(&epModelsWg, epModelsCh, eps[k])
 	}
-	endpointmanager.Mutex.RUnlock()
 
-	for i := 0; i < numEndpoints; i++ {
-		eps = append(eps, <-modelChan)
-	}
+	epsAppendWg.Add(1)
+	go func(epsAppended *sync.WaitGroup) {
+		for ep := range epModelsCh {
+			resEPs = append(resEPs, ep)
+		}
+		epsAppended.Done()
+	}(&epsAppendWg)
```

Comment:
Why are we doing this in a goroutine? we're waiting anyway so we can just do it in the Handle function directly, or is there something subtle I'm missing?

Since we have to return a slice as the payload we can be a little less elegant:
- instead of a channel, make a slice of the correct size
- In the functions getting the models, give them an index into the slice and have them write directly

I'm not sure how to handle nils in this scheme, however, so the channel might be the simplest.

In any case, though, I'm not convinced that the cost of copying some ints and strings around is so high that we have to do this in parallel. My assumption is that the getModel itself is nearly 0 cost but that the wait on locks may be higher, per endpoint. In that case, how long would the write locks be held? I think we only do it to regenerate or update things, but even then might not hold the lock itself for long. shrug. (this is a rant, not something to change in this PR).

Reply:
I've added this comment:
```
	// This needs to be done over channels since we might not receive all
	// the existing endpoints since not all endpoints contain the list of
	// labels that we will use to filter in `ep.HasLabels(convertedLabels)`
```